socrata/socrata_upload.py

```diff
@@ -0,0 +1,337 @@
+import contextlib
+import io
+import json
+import pandas as pd
+from pyathena import connect
+from pyathena.pandas.cursor import PandasCursor
+
+# Load environmental variables and connect to Athena
+app_token = os.getenv("SOCRATA_APP_TOKEN")
+auth = (os.getenv("SOCRATA_USERNAME"), os.getenv("SOCRATA_PASSWORD"))
+cursor = connect(
+    s3_staging_dir=str(os.getenv("AWS_ATHENA_S3_STAGING_DIR")) + "/",
+    region_name=os.getenv("AWS_REGION"),
+    cursor_class=PandasCursor,
+).cursor(unload=True)
+
+
+def get_asset_info(socrata_asset):
+    """
+    Simple helper function to retrieve asset-specific information from dbt.
+    """
+
+    os.chdir("./dbt")
+
+    DBT = dbtRunner()
+    dbt_list_args = [
+        "--quiet",
+        "list",
+        "--resource-types",
+        "model",
+        "--output",
+        "json",
+        "--output-keys",
+        "name",
+        "meta",
+    ]
+
+    print(f"> dbt {' '.join(dbt_list_args)}")
+    dbt_output = io.StringIO()
+    with contextlib.redirect_stdout(dbt_output):
+        DBT.invoke(dbt_list_args)
+
+    model = [
+        json.loads(model_dict_str)
+        for model_dict_str in dbt_output.getvalue().split("\n")
+        # Filter out empty strings caused by trailing newlines
+        if model_dict_str
+    ]
+
+    os.chdir("..")
+
+    model = pd.json_normalize(model)
+    model = model[model["meta.socrata_asset"] == socrata_asset]
+    athena_asset = model.iloc[0]["name"]
+    asset_id = model.iloc[0]["meta.asset_id"]
+    row_identifier = model.iloc[0]["meta.primary_key"]
+
+    return athena_asset, asset_id, row_identifier
+
+
+def build_query(athena_asset, row_identifier, years=None, township=None):
+    """
+    Build an Athena compatible SQL query. Function will append a year
+    conditional if `years` is non-empty. Many of the CCAO's open data assets are
+    too large to pass to Socrata without chunking. A `row_id` column is
+    constructed in order to use Socrata's `upsert` functionalities (updating
+    rather than overwriting data that already exists) based on the column names
+    passed to `row_indetifiers`.
+    """
+
+    row_identifier = "CONCAT(" + ", ".join(row_identifier) + ") AS row_id,"
+
+    # Retrieve column names and types from Athena
+    columns = cursor.execute("show columns from " + athena_asset).as_pandas()
+
+    # Array type columns are not compatible with the json format needed for
+    # Socrata uploads. Automatically convert any array type columns to string.
+    columns.loc[columns["type"] == "array(varchar)", "column"] = (
+        "ARRAY_JOIN("
+        + columns[columns["type"] == "array(varchar)"]["column"]
+        + ", ', ') AS "
+        + columns[columns["type"] == "array(varchar)"]["column"]
+    )
+
+    if not years:
+        query = (
+            "SELECT\n"
+            + row_identifier
+            + "\n"
+            + ",\n".join(columns["column"])
+            + "\nFROM "
+            + athena_asset
+            + "\nLIMIT 1000"
```

Comment:
**[Question, non-blocking]** I'm guessing these limits are for testing and we're going to eventually strip them out, is that right?

Reply:
Yep, I'll add a note to the PR. I also need to replace the test asset ID for parcel universe.